### PR TITLE
feat: extending openAPI

### PR DESCRIPTION
**Pull request by galzunited**
override the defaults with custom openAPI - self hosted css for future changes. as for now, refactored some of the apis and added new one so we could create user/events manually. excluded irrelevant apis from schema.

### CODE REVIEW

File: app/routers/dayview.py [hunk 1]
```diff
@@ -93,17 +92,17 @@ def _check_multiday_event(self) -> Tuple[bool]:
...
     # TODO: add a login session
     user = db_session.query(User).filter_by(username='test1').first()
     day = datetime.strptime(date, '%Y-%m-%d')
     day_end = day + timedelta(hours=24)
     events = db_session.query(Event).filter(
         Event.owner_id == user.id).filter(
-            or_(and_(Event.start >= day, Event.start < day_end),
-                and_(Event.end >= day, Event.end < day_end),
-                and_(Event.start < day_end, day_end < Event.end)))
+        or_(and_(Event.start >= day, Event.start < day_end),
```
Comment (yammesicka) [inline app/routers/dayview.py:103]: I actually prefer it in the previous way

File: app/utils/extending_openapi.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from fastapi.openapi.utils import get_openapi
+
+
+def custom_openapi(app):
+    if app.openapi_schema:
+        return app.openapi_schema
+    url = 'https://forums.pythonic.guru'
```
Comment (yammesicka) [inline app/utils/extending_openapi.py:7]: Prefer:
```
url = (
    'string1'
    'string2'
    'string3'
)
```

**Review by yammesicka — CHANGES_REQUESTED** [review]
Looks great! Please fix the conflicts and look at my 2 minor comments :)

File: app/routers/categories.py [hunk 3]
```diff
@@ -43,6 +43,19 @@ def get_categories(request: Request,
                                    f"unallowed params.")
 
 
+@router.get("/get_all_categories")
```
Comment (yammesicka) [inline app/routers/categories.py:46]: Prefer: `/categories` (if taken by another page, `/categories/list`)

File: app/routers/categories.py [hunk 4]
```diff
@@ -43,6 +43,19 @@ def get_categories(request: Request,
...
+def get_all_categories(db_session: Session = Depends(get_db)) -> List[
+                        Category]:
+    return db_session.query(Category).all()
+
+
+@router.post("/get_categories_by_user_id")
```
Comment (yammesicka) [inline app/routers/categories.py:52]: Change to `get`. The path should be `/categories?user_id=...`

File: app/routers/invitation.py [hunk 5]
```diff
@@ -56,9 +57,10 @@ def get_all_invitations(session: Session, **param) -> List[Invitation]:
         return invitations
 
 
+@router.post("/get_invitation_by_id")
 def get_invitation_by_id(
-        invitation_id: int, session: Session
-) -> Union[Invitation, None]:
+        invitation_id: int, session=Depends(get_db)) -> Union[
```
Comment (yammesicka) [inline app/routers/invitation.py:62]: Prefer splitting the lines in a more readable way :)

File: app/routers/user.py [hunk 6]
```diff
@@ -1,10 +1,42 @@
...
+
+
+class UserModel(BaseModel):
+    username: str
+    password: str
+    email: str = Field(regex='^\\S+@\\S+\\.\\S+$')
+    language: str
+
+
+@router.get("/get_all_users")
```
Comment (yammesicka) [inline app/routers/user.py:26]: `/user/`

File: app/routers/user.py [hunk 7]
```diff
@@ -1,10 +1,42 @@
...
+async def get_all_users(session=Depends(get_db)):
+    return session.query(User).all()
+
+
+@router.get("/get_user")
```
Comment (yammesicka) [inline app/routers/user.py:31]: `/user/<id>`

File: app/routers/user.py [hunk 8]
```diff
@@ -1,10 +1,42 @@
...
+async def get_user(user_id: int, session=Depends(get_db)):
+    return session.query(User).filter_by(id=user_id).first()
+
+
+@router.post("/create_user")
```
Comment (yammesicka) [inline app/routers/user.py:36]: `/user` (using `post`)

**Review by yammesicka — COMMENTED** [review]
Great progress, I've added few additional notes :)

File: app/routers/profile.py [hunk 9]
```diff
@@ -153,3 +168,71 @@ def get_image_crop_area(width, height):
...
+    try:
+        save_holidays_to_db(holidays, session)
+    except SQLAlchemyError as ex:
+        logger.exception(ex)
+    finally:
+        url = router.url_path_for("profile")
+        return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+def get_holidays_from_file(file: List[Event], session: Session) -> List[Event]:
```
Comment (yammesicka) [inline app/routers/profile.py:187]: I think these functions should be in the `internal` directory. WDYT?

File: app/routers/event.py [hunk 10]
```diff
@@ -36,13 +37,39 @@
...
+    owner_id: int
+    location: str
+
+
+@router.get("/get_events")
+async def get_events(session=Depends(get_db)):
+    return session.query(Event).all()
+
+
+@router.post("/create_event_manually")
```
Comment (yammesicka) [inline app/routers/event.py:54]: Why not using the regular create event route?

File: app/routers/event.py [hunk 11]
```diff
@@ -36,13 +37,39 @@
...
+class EventModel(BaseModel):
+    title: str
+    start: dt
+    end: dt
+    content: str
+    owner_id: int
+    location: str
+
+
+@router.get("/get_events")
```
Comment (yammesicka) [inline app/routers/event.py:49]: `/events`

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/routers/profile.py (hunk 9 above)
Comment (galzunited) [reply app/routers/profile.py:187]: i think you right. but it is related to my second feature so ill fix it there.

File: app/routers/categories.py [hunk 12]
```diff
@@ -43,6 +43,19 @@ def get_categories(request: Request,
...
+@router.get("/list")
+def get_all_categories(db_session: Session = Depends(get_db)) -> List[
+                        Category]:
+    return db_session.query(Category).all()
+
+
+@router.get("/")
+def get_categories_by_user_id(user_id: int,
+                              db_session: Session = Depends(get_db)) -> List[
+                              Category]:
```
Comment (yammesicka) [inline app/routers/categories.py:55]: This one is harder to read than:

```python
def get_categories_by_user_id(
    user_id: int, db_session: Session = Depends(get_db),
) -> List[Category]:
```

File: app/routers/event.py [hunk 13]
```diff
@@ -37,13 +38,39 @@
...
+@router.post("/create_event")
+async def create_event_api(event: EventModel, session=Depends(get_db)):
+    create_event(db=session,
+                 title=event.title,
+                 start=event.start,
+                 end=event.start,
+                 content=event.content,
+                 owner_id=event.owner_id,
+                 location=event.location)
+    return f'Successfully created {event.title} event'
```
Comment (yammesicka) [inline app/routers/event.py:64]: Prefer returning a json: `{'success': True}` (if you want it to be perfect, return the details of the created event)

File: app/routers/event.py [hunk 14]
```diff
@@ -37,13 +38,39 @@
...
+    owner_id: int
+    location: str
+
+
+@router.get("/events")
+async def get_events(session=Depends(get_db)):
+    return session.query(Event).all()
+
+
+@router.post("/create_event")
```
Comment (yammesicka) [inline app/routers/event.py:55]: Just `/` with `post`

File: app/routers/event.py [hunk 15]
```diff
@@ -37,13 +38,39 @@
...
+class EventModel(BaseModel):
+    title: str
+    start: dt
+    end: dt
+    content: str
+    owner_id: int
+    location: str
+
+
+@router.get("/events")
```
Comment (yammesicka) [inline app/routers/event.py:50]: Just `/` (you already have `/event` as the root of this API)

File: app/routers/whatsapp.py [hunk 16]
```diff
@@ -1,9 +1,9 @@
 from typing import Optional
-from urllib.parse import urlencode
 
 from fastapi import APIRouter
+from urllib.parse import urlencode
 
-router = APIRouter()
+router = APIRouter(tags=["utis"])
```
Comment (yammesicka) [inline app/routers/whatsapp.py:6]: Maybe `utils`?

File: app/templates/profile.html [hunk 17]
```diff
@@ -196,7 +196,39 @@ <h6 class="card-title text-center mb-1">{{ user.full_name }}</h6>
...
+                        </ul>
+                    </div>
+                </div>
+                <!-- End Features card -->
+
+            </div>
+
+            <!-- Center -->
+            <div class="col-5">
+                    </p>
```
Comment (yammesicka) [inline app/templates/profile.html:231]: What is the purpose of the `</p>` tag here?

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/templates/profile.html (hunk 17 above)
Comment (galzunited) [reply app/templates/profile.html:231]: thats not mine :D removing

**Comment by yammesicka** [comment]
Please solve the conflicts :)

File: app/routers/event.py [hunk 18]
```diff
@@ -38,13 +39,39 @@
...
+async def create_event_api(event: EventModel, session=Depends(get_db)):
+    create_event(db=session,
+                 title=event.title,
+                 start=event.start,
+                 end=event.start,
+                 content=event.content,
+                 owner_id=event.owner_id,
+                 location=event.location)
+    return {'success': True}
+
```
Comment (yammesicka) [inline app/routers/event.py:66]: Where are the tests for these function? :)

**Review by yammesicka — COMMENTED** [review]
Almost there :)

**Review by yammesicka — APPROVED** [review]

**Comment by yammesicka** [comment]
I've fixed some conflicts in order to merge, but there are some errors - please have a look

https://github.com/PythonFreeCourse/calendar/pull/226/checks?check_run_id=1896171413

File: app/routers/profile.py [hunk 19]
```diff
@@ -164,7 +164,7 @@ def get_image_crop_area(width, height):
     return 0, delta, width, width + delta
 
 
-@router.post("/holidays/update")
+@router.post("/holidays/update_holidays")
```
Comment (yammesicka) [inline app/routers/profile.py:167]: Just `/update`

File: app/templates/import_holidays.html [hunk 20]
```diff
@@ -9,7 +9,7 @@ <h1 class="title">Import holidays using ics file</h1>
     <a href="https://www.feiertagskalender.ch/export.php?geo=3542&hl=en#" target="_blank">Check this website to export holidays by country</a>
     <div class="input-group">
             <label for="file"></label>
```
Comment (yammesicka) [inline app/templates/import_holidays.html:11]: Preferably, add text inside the `label`

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great job. I have some minor insights. This issue is ready to merge after fixing them.

File: app/static/style.css [hunk 21]
```diff
@@ -96,14 +96,17 @@ p {
...
 }
 
+.input-upload-file {
+    margin-top: 15px
+}
 
-.subtitle {
-    font-size: 20px;
+.upload-file {
+    margin: auto 10px auto 0px
```
Comment (yammesicka) [inline app/static/style.css:111]: Use `em`s instead and add `;` in the end of the line

File: app/static/style.css [hunk 22]
```diff
@@ -96,14 +96,17 @@ p {
...
-  
+
 .error-message {
     line-height: 0;
     color: red;
     padding-left: 12.5rem;
 }
 
+.input-upload-file {
+    margin-top: 15px
```
Comment (yammesicka) [inline app/static/style.css:107]: Use `em` instead and add `;` at the end of the line

**Review by yammesicka — CHANGES_REQUESTED** [review]

**Pull request merged**